lambda/src/main/scala/pricemigrationengine/migrations/Newspaper2024Migration/Newspaper2024MigrationEstimation.scala

```diff
@@ -0,0 +1,253 @@
+          billingPeriod = billingPeriod,
+          currency = currency,
+          currentPrice = currentPrice,
+          targetRatePlanId = targetRatePlanId
+        )) match {
+          case Some(data) => Right(data)
+          case _ => Left(s"[error: 0e218c37] Could not determine SubscriptionData2024 for subscription ${subscription}")
+        }
+      }
+      case _ =>
+        Left(
+          s"[error 93a21a48] Subscription ${subscription.subscriptionNumber} was found to have more than one newsPaperDeliveryRatePlans making determination of rate plan name impossible"
+        )
+    }
+  }
+
+  /*
+
+    case class SubscriptionData2024(
+      productName: String,
+      ratePlan: ZuoraRatePlan,
+      ratePlanName: String,
+      billingPeriod: BillingPeriod,
+      currency: String,
+      currentPrice: BigDecimal
+  )
+   */
+
+  def subscriptionToNewPrice(subscription: ZuoraSubscription): Option[BigDecimal] = {
+    // The price correction factor is meant to be 1 except in the case of
+    // subscriptions with capping in which case the correction factor will have been precomputed
+    // from the data generated by Matt
+    val priceCorrectionFactor = Newspaper2024MigrationPriceCapping.priceCorrectionFactor(subscription)
+    for {
+      data2024 <- subscriptionToSubscriptionData2024(subscription).toOption
+      price <- Newspaper2024MigrationStaticData.priceLookup(
+        data2024.productName,
+        data2024.billingPeriod,
+        data2024.ratePlanName
+      )
+    } yield price * priceCorrectionFactor
+  }
+
+  def priceData(
+      subscription: ZuoraSubscription,
+  ): Either[AmendmentDataFailure, PriceData] = {
+
+    // PriceData(currency: Currency, oldPrice: BigDecimal, newPrice: BigDecimal, billingPeriod: String)
+
+    def transform1[T](option: Option[T]): Either[AmendmentDataFailure, T] = {
+      option match {
+        case None                 => Left(AmendmentDataFailure("error"))
+        case Some(ratePlanCharge) => Right(ratePlanCharge)
+      }
+    }
+
+    def transform2[T](data: Either[String, T]): Either[AmendmentDataFailure, T] = {
+      data match {
+        case Left(string) => Left(AmendmentDataFailure(string))
+        case Right(t)     => Right(t)
+      }
+    }
+
+    for {
+      data2024 <- transform2[SubscriptionData2024](subscriptionToSubscriptionData2024(subscription))
+      oldPrice = data2024.currentPrice
+      newPrice <- transform1[BigDecimal](subscriptionToNewPrice(subscription))
+    } yield PriceData(
+      data2024.currency,
+      oldPrice,
+      newPrice,
+      BillingPeriod.toString(data2024.billingPeriod)
+    )
+  }
+
+  def subscriptionToBatchId(subscription: ZuoraSubscription): Either[String, Newspaper2024BatchId] = {
+    val ratePlanDetails = (for {
+      ratePlanDetails <- subscriptionToSubscriptionData2024(subscription)
+    } yield ratePlanDetails)
+
+    ratePlanDetails match {
+      case Left(message) => Left(message)
+      case Right(ratePlanDetails) => {
```

Comment:
```suggestion
    ratePlanDetails.map {
      details => {
```

Reply:
Here that change would not work as well because the `Right` case can still return a left due to a further check. So going to leave it as a `match` instead of refactoring into a `map` :)

Reply:
On second thought I could still see that those indented checks were not really idiomatic Scala either (no pun intended), and the solution was simply to not exit the `for` construct as I did in the original version. The refactoring that I pushed here ( https://github.com/guardian/price-migration-engine/pull/964/commits/d22540a8046a9ba0967742340c4dcbf84160ee7b ) solves it.